test/test_device_histogram.cu

```diff
@@ -426,77 +426,72 @@ struct SearchTransform
     }
 };
 
-
 // Scales samples to evenly-spaced bins
 template <typename LevelT>
 struct ScaleTransform
 {
-    int    num_levels;  // Number of levels in array
-    LevelT max;         // Max sample level (exclusive)
-    LevelT min;         // Min sample level (inclusive)
-    LevelT scale;       // Bin scaling factor
-
-    void Init(
-        int    num_levels_,  // Number of levels in array
-        LevelT max_,         // Max sample level (exclusive)
-        LevelT min_,         // Min sample level (inclusive)
-        LevelT scale_)       // Bin scaling factor
-    {
-        this->num_levels = num_levels_;
-        this->max = max_;
-        this->min = min_;
-        this->scale = scale_;
-    }
+  int num_levels; // Number of levels in array
+  LevelT max;     // Max sample level (exclusive)
+  LevelT min;     // Min sample level (inclusive)
 
-    // Functor for converting samples to bin-ids  (num_levels is returned if sample is out of range)
-    template <typename SampleT>
-    int operator()(SampleT sample)
-    {
-        if ((sample < min) || (sample >= max))
-        {
-            // Sample out of range
-            return num_levels;
-        }
+  void Init(int num_levels_, // Number of levels in array
+            LevelT max_,     // Max sample level (exclusive)
+            LevelT min_)     // Min sample level (inclusive)
+  {
+    this->num_levels = num_levels_;
+    this->max        = max_;
+    this->min        = min_;
+  }
 
-        return (int) ((((LevelT) sample) - min) / scale);
+  // Functor for converting samples to bin-ids  (num_levels is returned if sample is out of range)
+  template <typename SampleT>
+  int operator()(SampleT sample)
+  {
+    if ((sample < min) || (sample >= max))
+    {
+      // Sample out of range
+      return num_levels;
     }
+
+    // Accurate bin computation following the arithmetic we guarantee in the HistoEven docs
+    return static_cast<int>(
+      (static_cast<uint64_t>(sample - min) * static_cast<uint64_t>(num_levels - 1)) /
```

Comment:
Good point! 👍 I've added a specialisation for `half_t` that keeps the current behaviour for bin computation.

Reply:
I'd prefer a bit different approach. I'd leave existing version as primary template, while specializing for integral and floating types. What do you think?

```cpp
// primary template
template <typename LevelT, typename = void>
struct ScaleTransform;

// integers
template <typename LevelT>
struct ScaleTransform<LevelT, 
                      typename ::cuda::std::enable_if<::cuda::std::is_integral<LevelT>::value>::type>;

// fp
template <typename LevelT>
struct ScaleTransform<LevelT, 
                      typename ::cuda::std::enable_if<
                            ::cuda::std::is_floating_point<LevelT>::value 
                            #if TEST_HALF_T
                        || ::cuda::std::is_same<LevelT, half_t>::value
                            #endif
                        >::type
                      >;
```

Reply:
Adopted your suggestion but also had to change the logic for computing the scale and bins for `half_t`, because `half_t` were previously (i.e., before this PR) covered by the `ScaleTransform` template that was also specialised for integral types.